### PR TITLE
don't require .git suffix; consolidate validation, empty protocol URL processing

### PR DESCRIPTION
**Pull request by gabemontero**
@bparees @csrwng from email exchange this weekend and @csrwng heads up on Friday.  Also added a quick verification for bugzilla 1268478

### CODE REVIEW

File: pkg/scm/git/git.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ type stiGit struct {
 	runner util.CommandRunner
 }
 
-var gitSshURLExp = regexp.MustCompile(`\A([\w\d\-_\.+]+@[\w\d\-_\.+]+:[\w\d\-_\.+%/]+\.git)$`)
+var gitSshURLExp = regexp.MustCompile(`\A([\w\d\-_\.+]+@[\w\d\-_\.+]+:[\w\d\-_\.+%/])`)
```
Comment (csrwng) [inline pkg/scm/git/git.go]: Is there a reason you got rid of the '$' at the end?

Comment (gabemontero) [inline pkg/scm/git/git.go]: the test cases did not pass with it present with with \.git removed

**Comment by gabemontero** [comment]
[test]

File: pkg/scm/git/git.go (hunk 1 above)
Comment (csrwng) [inline pkg/scm/git/git.go]: If I add ".git" to "git@192.168.122.1:repositories/authooks" (as in "git@192.168.122.1:repositories/authooks.git"), the test case passes (with the "$" at the end). I think it needs some more investigation. Also, I'm not sure that we're doing the best validation with the current string. For example, we would allow a username that starts with "-" or "+" which I think is invalid.

Comment (gabemontero) [inline pkg/scm/git/git.go]: But the point of that particular test was to allow URLs without .git at the end.  If I add $, then that URL without .git fails.

Comment (csrwng) [inline pkg/scm/git/git.go]: My point is that it's not clear why it fails

Comment (csrwng) [inline pkg/scm/git/git.go]: If you don't include "$" at the end, then you can have any chars at the end of the string and they don't have to match your regex... like gabe@host:/test!@#.\***blah

Comment (csrwng) [inline pkg/scm/git/git.go]: Also, while you investigate, don't forget about IPv6:
http://www.chtaube.eu/news/2014/0918-git210-parser-bug-in-ipv6-ssh-urls/

Comment (csrwng) [inline pkg/scm/git/git.go]: Also specifying a custom port:
http://stackoverflow.com/questions/5767850/git-on-custom-ssh-port

Comment (gabemontero) [inline pkg/scm/git/git.go]: Note on [custom port](http://stackoverflow.com/questions/5767850/git-on-custom-ssh-port), the commentary there explicitly states that you can only specify a port *IF* you specify the protocol (man git-clone also implies this).  As such, golang's url.Parse handles the strings where the protocol is specified, so we should be good there.

**Comment by gabemontero** [comment]
@csrwng @bparees I've pushed the consolidation of git url validation / munging when no protocol is specified; included are a couple of new methods on the interface which are leveraged by both source-to-image and origin

**Comment by gabemontero** [comment]
@csrwng @bparees actually wait ... I thought I saw hack/test-go.sh passing but now it is not

**Comment by gabemontero** [comment]
@csrwng @bparees ok now it is good (all hack/test-*.sh and relevant hack/verify-*.sh passing) - ptal

File: pkg/scm/git/git.go [hunk 2]
```diff
@@ -62,18 +68,107 @@ func stringInSlice(s string, slice []string) bool {
...
+func validCloneSpec(source string, remoteOnly bool, partialURL *url.URL) bool {
+	if partialURL == nil {
+		uri, err := url.Parse(source)
+		if err != nil {
+			return false
+		}
+		partialURL = uri
+	}
+
+	if strings.Index(source, "~") != -1 {
```
Comment (csrwng) [inline pkg/scm/git/git.go]: Is there a reason you explicitly test for the '~'?

Comment (gabemontero) [inline pkg/scm/git/git.go]: To maintain previous behavior for this test:  https://github.com/openshift/origin/blob/master/pkg/generate/app/cmd/newapp_test.go#L35-L97

The ~ stuff with new-app we talked about on Monday.

Comment (csrwng) [inline pkg/scm/git/git.go]: I guess I don't understand why you check for ~ explicitly... wouldn't it be detected by the regex's for the individual pieces?

Comment (gabemontero) [inline pkg/scm/git/git.go]: It wasn't originally, but some of my changes to the regex's may have occurred after I added that check.  I'll try again now.

Comment (gabemontero) [inline pkg/scm/git/git.go]: Yep ... it works fine now with the latest regex's and that check removed.  I'll update the PR.

**Comment by gabemontero** [comment]
[test]

File: pkg/scm/git/git.go [hunk 3]
```diff
@@ -62,18 +68,103 @@ func stringInSlice(s string, slice []string) bool {
...
+				return false
+			}
+		}
+		partialURL.Scheme = "ssh"
+		partialURL.User = url.User(user)
+		partialURL.Host = host
+		partialURL.Path = path
+		partialURL.Fragment = ref
+		return true
+	} else if !remoteOnly {
```
Comment (csrwng) [inline pkg/scm/git/git.go]: you can have a valid remote clone spec without a user:
git clone linux-box:/data/openshift-docs works fine on my Mac

File: pkg/scm/git/git_test.go [hunk 4]
```diff
@@ -10,7 +10,7 @@ import (
 )
```
Comment (csrwng) [inline pkg/scm/git/git_test.go]: You should add a set of tests for ValidCloneSpecRemoteOnly

File: pkg/scm/git/git.go (hunk 3 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: If the user string is empty, we end up with the same empty Username struct that url.Parse(string) started us on.  My take was that a simple re-initialization was cleaner that and if len(user) > 0 check.

Based on that clarification, would you agree?

File: pkg/scm/git/git_test.go (hunk 4 above)
Comment (gabemontero) [inline pkg/scm/git/git_test.go]: Yep - missed that.  I'll add some.

**Comment by gabemontero** [comment]
ok, pushed changes resulting from some face to face back and forth between @csrwng and I, as well as some additional fixes in scm.go/scm_test.go based on similar errors on what is a valid git clone spec

**Comment by gabemontero** [comment]
@csrwng @bparees woke up this AM and a couple of minor tweaks came to mind

ptal at your earliest convenience

File: pkg/scm/git/git.go [hunk 5]
```diff
@@ -62,18 +67,136 @@ func stringInSlice(s string, slice []string) bool {
...
+// corrections to the URL when no protocol is specified in the Scheme
+func (h *stiGit) MungeNoProtocolURL(source string, uri *url.URL) bool {
+	if uri == nil {
+		return false
+	}
+	rc, scheme, user, host, path, ref := validCloneSpec(source, false, uri)
+	if !rc {
+		return false
+	}
+	if scheme != "ssh" && scheme != "file" {
```
Comment (bparees) [inline pkg/scm/git/git.go]: this is confusing behavior....  at a minimum i think it needs some comments.

if the uri wasn't a validCloneSpec, the function returns false.  but it also returns false if it was valid, but the protocol was anything other than "file" or "ssh"?  (which is an odd pairing of protocols to be treated the same)

File: pkg/test/git.go [hunk 6]
```diff
@@ -24,6 +25,18 @@ func (f *FakeGit) ValidCloneSpec(source string) bool {
 	return f.ValidCloneSpecResult
 }
 
+//ValidCloneSpecRemoteOmly returns a valid GIT clone specification
```
Comment (bparees) [inline pkg/test/git.go]: s/Omly/Only/

File: pkg/scm/scm.go [hunk 7]
```diff
@@ -16,41 +15,48 @@ import (
...
 
-	// Convert relative path to absolute path.
-	if !strings.HasPrefix(s, "/") {
-		if absolutePath, err := filepath.Abs(s); err == nil {
-			s = absolutePath
+	if fileCopy {
+		if doesExist(s, true) {
+			if !isLocalGitRepository(s) {
+				// since not using git, any resulting URLs need to be explicit with file:// protocol specified
+				s = "file://" + s
```
Comment (bparees) [inline pkg/scm/scm.go]: what if the user provided the path with a file:// prefix already on it?

File: pkg/scm/scm.go [hunk 8]
```diff
@@ -16,41 +15,48 @@ import (
...
-	// If we have local directory and that directory exists, use file copy
-	if _, err := os.Stat(s); err == nil {
-		return &file.File{util.NewFileSystem()}, s, nil
+	// If the source is valid  GIT protocol (file://, git://, git@, etc..) use GIT
+	// binary to download the sources
+	if g.ValidCloneSpec(s) {
+		// if local file system, without file://, when using git, should not need file://, but we'll be safe;
+		// satisfies previous constructed test case in scm_test.go as well
+		if doesExist(s, false) {
+			s = "file://" + s
```
Comment (bparees) [inline pkg/scm/scm.go]: again what if they provided the path with file:// on it?

File: pkg/scm/git/git.go [hunk 9]
```diff
@@ -18,6 +18,8 @@ import (
 // Git is an interface used by main STI code to extract/checkout git repositories
 type Git interface {
 	ValidCloneSpec(source string) bool
+	ValidCloneSpecRemoteOnly(source string) bool
+	MungeNoProtocolURL(source string, url *url.URL) bool
```
Comment (bparees) [inline pkg/scm/git/git.go]: it looks like this is never called, what am i missing?

Comment (gabemontero) [inline pkg/scm/git/git.go]: MungeNoProtocolURL will be called by the pending origin changes.  There was duplicate logic in origin modules that have been consolidated in pkg/scm/git/git.go

File: pkg/scm/scm.go (hunk 7 above)
Comment (gabemontero) [inline pkg/scm/scm.go]: ah ...late night coding ... the boolean param on doesExist was part of an attempt to deal with file:// either being there or not, without multiple checks for the prefix, but did not complete the job;

i'll rework and update

File: pkg/scm/scm.go (hunk 8 above)
Comment (gabemontero) [inline pkg/scm/scm.go]: see comment above

File: pkg/scm/git/git.go (hunk 5 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: yeah, the boolean return was in fact a placeholder for future use - its actual use in my corresponding origin change does not leverage it... certainly with lack of use I did not think it through.

I'm inclined to switch gears and have a void return.  Will do that as part of next push.

On comments, yeah, I had some on the origin side but not here.  I'll address.

On the file and ssh pairing, just to be clear, there "association" such as it is stems from the fact that from a git clone spec perspective, there are valid url's where neither file:// or ssh:// are specified, but based on other characters being present (or not present) the file:// or ssh:// can be derived.

**Comment by gabemontero** [comment]
@bparees @csrwng updates in response to @bparees comments pushed

File: pkg/scm/git/git.go [hunk 10]
```diff
@@ -62,18 +69,145 @@ func stringInSlice(s string, slice []string) bool {
...
+		shouldReturn, returnCode := schemeNecessitatesReturn(partialURL)
+		if shouldReturn {
+			return returnCode, scheme, user, host, path, ref
+		}
 	}
 
-	if stringInSlice(url.Scheme, allowedSchemes) {
-		return true
+	lastColonIdx := strings.LastIndex(source, ":")
+	if lastColonIdx != -1 {
```
Comment (bparees) [inline pkg/scm/git/git.go]: can't this ":" be part of "file:///some/local/path" rather than part of an ssh reference?

Comment (bparees) [inline pkg/scm/git/git.go]: oh i guess that would have been caught earlier by the "schemeNecessitatesReturn" check.

Comment (bparees) [inline pkg/scm/git/git.go]: definitely calls for a "// at this point we have elimated all source paths that have an explicit protocol, now we need to distinguish between a local file path that wasn't prefixed with file:// and an ssh path that wasn't prefixed with ssh://" comment.

Comment (bparees) [inline pkg/scm/git/git.go]: also "colon" is  a valid character in a filename, so is this really sufficient?

File: pkg/scm/git/git.go [hunk 11]
```diff
@@ -62,18 +69,145 @@ func stringInSlice(s string, slice []string) bool {
...
+	return false, true
+}
+
+func validCloneSpec(source string, remoteOnly bool, partialURL *url.URL) (rc bool, scheme, user, host, path, ref string) {
+	// note, as part of parsing the clone spec to see if it is valid, various values for the fields in url.URL are saved
+	// and returned for possible URL corrections from MungeNoProtocolURL
+	if partialURL == nil {
+		uri, err := url.Parse(source)
+		if err != nil {
+			return false, scheme, user, host, path, ref
```
Comment (csrwng) [inline pkg/scm/git/git.go]: When you have named return parameters, you can just set the value of each var and just do a "return" with no arguments. It also seems to me that what you are returning could be expressed in a URL struct.

File: pkg/scm/git/git.go [hunk 12]
```diff
@@ -62,18 +69,145 @@ func stringInSlice(s string, slice []string) bool {
...
+		uri.User = url.User(user)
+	}
+	uri.Scheme = scheme
+	uri.Host = host
+	uri.Path = path
+	uri.Fragment = ref
+	return nil
+}
+
+func schemeNecessitatesReturn(uri *url.URL) (shouldReturn, returnCode bool) {
```
Comment (csrwng) [inline pkg/scm/git/git.go]: It's not clear from the name of the function what it does or what the return values mean. At the very least you should have a comment describing what it does.

File: pkg/scm/git/git.go [hunk 13]
```diff
@@ -62,18 +69,145 @@ func stringInSlice(s string, slice []string) bool {
...
+	// note, there also appears to be a bug in url.Parse, where a url like
+	// linux-box:/path/to/code will report the scheme as "linux-box"
+	if uri.Scheme != "" {
+		return true, false
+	}
+
+	return false, true
+}
+
+func validCloneSpec(source string, remoteOnly bool, partialURL *url.URL) (rc bool, scheme, user, host, path, ref string) {
```
Comment (csrwng) [inline pkg/scm/git/git.go]: I think you can do without the partialURL parameter and call this parseSpec()... let the MungeNoProtocol function call it if it needs to.

File: pkg/scm/git/git.go (hunk 10 above)
Comment (csrwng) [inline pkg/scm/git/git.go]: a colon is not valid in a filename in a git clone spec

File: pkg/scm/git/git.go (hunk 11 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: true (return no args) ... I consciously added the arguments on the return because I just like things explicit 

true (a URL struct) .... my thinking was that if this method was public and consumed outside of this go file, then formalize the return with a struct.  But since it is only used locally. it wasn't necessary.

Given those rationales, how strong is your opinion here?  Still change it?

Comment (csrwng) [inline pkg/scm/git/git.go]: hmm ... for return no args... I don't know that explicitly including all vars makes the code clearer. For example, in the line above you really haven't set any values to scheme, user, host,  etc. Setting rc to false and then a return might make it clearer what it is you're itending to do. 

BTW, rc doesn't tell me what it means... valid might be a better name?

I guess as far as the struct goes, I guess it's a matter of preference. A struct would make it clearer for me that your return has 2 parts... one (the bool) indicating whether the input is valid or not and the other one breaking down the parts of the spec.

Comment (gabemontero) [inline pkg/scm/git/git.go]: Ah ... the "2 parts" thing is a good point.  Ok, I've been converted on the struct point.  And sure, I'll come up with something more specific than "rc".

On the return no args, I suppose it comes down to where you draw the line with being explicit/implicit ... i.e. the implicit setting of string vars to "" ;-) ... I'll revisit after the struct conversion and see if I want to continue the discussion on that point then. Thx.

File: pkg/scm/git/git.go (hunk 12 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: added additional comments for now ... this function as it currently exists might change (or even go away) based on the outcome of the next comment.

File: pkg/scm/git/git.go (hunk 13 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: yep - i like it - thx - part of next push

File: pkg/scm/git/git.go (hunk 10 above)
Comment (bparees) [inline pkg/scm/git/git.go]: so what's going to happen if i run:  "s2i build some/lo:cal/path"?

Comment (gabemontero) [inline pkg/scm/git/git.go]: just added that to the test cases - it is correctly identified as an invalid git clone spec

Comment (bparees) [inline pkg/scm/git/git.go]: but it should be accepted as a valid source path because it's a local file system path, is it?

Comment (gabemontero) [inline pkg/scm/git/git.go]: Ah - I see now.  Yep, you are right.

And y'all are probably aware, but just in case, today the s2i code does a golang os level copy of files for either file:// or just plain paths.

So to me it would make sense to allow ":" in both those cases.  Agreed?

The trick will be distinguishing between protocol-less file spec and ssh spec.

Comment (gabemontero) [inline pkg/scm/git/git.go]: FYI - this looks like the description of what is valid for git specs:  https://git-scm.com/docs/git-check-ref-format

Definitely more restrictive that what is allowed in URLs:  http://stackoverflow.com/questions/2849756/list-of-valid-characters-for-the-fragment-identifier-in-an-url

**Comment by gabemontero** [comment]
@bparees @csrwng updates from this AM's comments have been pushed

File: pkg/scm/git/git.go [hunk 14]
```diff
@@ -62,18 +78,178 @@ func stringInSlice(s string, slice []string) bool {
...
+	if len(mods.path) > 0 {
+		uri.Path = mods.path
+	}
+	if len(mods.frag) > 0 {
+		uri.Fragment = mods.frag
+	}
+	return nil
+}
+
+// schemeNeccessitatesReturn examines the URL and determines whether further
```
Comment (bparees) [inline pkg/scm/git/git.go]: schemeNecessitatesReturn

File: pkg/scm/git/git.go [hunk 15]
```diff
@@ -62,18 +78,178 @@ func stringInSlice(s string, slice []string) bool {
...
+// url / git spec string processing is needed or not, and if no more processing is needed,
+// whether the input should be deemed valid or not;
+// the combos of interest are:
+// - if a valid protocol/scheme is specifed, just return (leaving the valid URL alone)
+// - if an invalid protocol/scheme is specified
```
Comment (bparees) [inline pkg/scm/git/git.go]: do what?

File: pkg/scm/git/git.go [hunk 16]
```diff
@@ -62,18 +78,178 @@ func stringInSlice(s string, slice []string) bool {
...
+			}
+		}
+
+		if !gitSshURLIPv6.MatchString(host) && !gitSshURLIPv4.MatchString(host) {
+			return nil, errors.New(fmt.Sprintf("invalid host provided: %s from %s", host, source))
+		}
+
+		segments = strings.SplitN(path, "#", 2)
+		if len(segments) == 2 {
+			path, frag = segments[0], segments[1]
```
Comment (bparees) [inline pkg/scm/git/git.go]: why the change from "ref" to "frag" as the name here?

File: pkg/scm/git/git.go [hunk 17]
```diff
@@ -62,18 +78,178 @@ func stringInSlice(s string, slice []string) bool {
...
+	if invalidProtocol {
+		return nil, errors.New(fmt.Sprintf("unsupported protocol specfied:  %s", uri.Scheme))
+	}
+	if justReturn {
+		return nil, nil
+	}
+
+	lastColonIdx := strings.LastIndex(source, ":")
+	atSignPresent := strings.Index(source, "@") != -1
+	if lastColonIdx != -1 && atSignPresent {
```
Comment (bparees) [inline pkg/scm/git/git.go]: not to make you feel like i'm going back and forth on you, but this would imply that:
github.com:openshit/origin.git  would not be a valid ssh clone spec, but it appears that it is:
```
$ git clone user@server:project.git

You can also not specify a user, and Git assumes the user you’re currently logged in as.
```
(from https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols)

at the end of the day there may not be any valid way to distinguish between a protocol-less ssh spec, and a protocoless local file spec in which the path includes a colon.

I wonder how git distinguishes, since git does allow local file paths as clone specs too.  Perhaps it just tries the local path first and if it doesn't exist, tries it as an ssh spec.  Presumably we could do the same thing here...  i know you already have some logic that checks if the path provided exists locally, maybe if it does you should just immediately augment the path with file:// (if it doesn't already have that prefix) and then you can avoid a lot of this nastiness.

File: pkg/scm/git/git.go (hunk 16 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: cause in the url doc the describe the piece after the # as a fragment

http://stackoverflow.com/questions/2849756/list-of-valid-characters-for-the-fragment-identifier-in-an-url
http://tools.ietf.org/html/rfc3986

File: pkg/scm/git/git.go (hunk 14 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: i need new glasses :-)

File: pkg/scm/git/git.go (hunk 15 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: comment update in hand

File: pkg/scm/git/git.go (hunk 16 above)
Comment (bparees) [inline pkg/scm/git/git.go]: yeah but this error message "invalid fragment provided:" is likely going to be confusing to users.  per our syntax it's a ref, because we're treating the value of that fragment as a git ref.

File: pkg/scm/git/git.go (hunk 17 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: yeah, @csrwng and I ran into this a day or two ago, and we decided not to support this (presumably) little used form given the ugliness of the fix.

I'll give it a go and see if I can get a solution that is easy enough on the stomach.

**Comment by gabemontero** [comment]
@bparees @csrwng ptal

**Comment by gabemontero** [comment]
missed the comeback on ref/frag ... i'll switch to ref

**Comment by gabemontero** [comment]
@bparees @csrwng ptal

File: pkg/scm/git/git.go [hunk 18]
```diff
@@ -62,18 +78,199 @@ func stringInSlice(s string, slice []string) bool {
...
+	}
+
+	lastColonIdx := strings.LastIndex(source, ":")
+	atSignPresent := strings.Index(source, "@") != -1
+	// deal with the special git case where a user is not specified (i.e. no @);
+	// to do so, see if the entire spec as is exists as a file
+	trySSH := false
+	if lastColonIdx != -1 && !atSignPresent {
+		path := source
+		hashLoc := strings.LastIndex(source, "#")
```
Comment (bparees) [inline pkg/scm/git/git.go]: filenames can contain # too.  i'd just check the source exactly as is to see if it's a valid local path.

File: pkg/scm/git/git.go [hunk 19]
```diff
@@ -62,18 +78,199 @@ func stringInSlice(s string, slice []string) bool {
...
+	if justReturn {
+		return nil, nil
+	}
+
+	lastColonIdx := strings.LastIndex(source, ":")
+	atSignPresent := strings.Index(source, "@") != -1
+	// deal with the special git case where a user is not specified (i.e. no @);
+	// to do so, see if the entire spec as is exists as a file
+	trySSH := false
+	if lastColonIdx != -1 && !atSignPresent {
```
Comment (bparees) [inline pkg/scm/git/git.go]: && !remoteOnly  ?

File: pkg/scm/git/git.go (hunk 18 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: ok - change coming soon

File: pkg/scm/git/git.go (hunk 19 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: yep - change coming soon

**Comment by gabemontero** [comment]
ok .. with change re: the exchange between @bparees and I last night - @bparees @csrwng ptal

**Comment by gabemontero** [comment]
@bparees @csrwng the latest refactor from this AM's scrum has been pushed to the PR

File: pkg/scm/git/git.go [hunk 20]
```diff
@@ -62,18 +84,272 @@ func stringInSlice(s string, slice []string) bool {
...
+func ParseURL(source string) (*url.URL, error) {
+	uri, err := url.Parse(source)
+	if err != nil {
+		return nil, err
+	}
+
+	// to work around the url.Parse bug, deal with the fact it seemed to split on just :
+	if strings.Index(source, "://") == -1 && uri.Scheme != "" {
+		uri.Scheme = ""
+		return nil, nil
```
Comment (bparees) [inline pkg/scm/git/git.go]: should this not return an error also?  (that it's not a valid url with a protocol)

File: pkg/scm/git/git.go [hunk 21]
```diff
@@ -62,18 +84,272 @@ func stringInSlice(s string, slice []string) bool {
...
+// expect git clone spec syntax; it also provides details if the file://
+// proto was explicitly specified, if we should use OS copy vs. the git
+// binary, and if a frag/ref has a bad format
+func ParseFile(source string) (details *FileProtoDetails, mods *URLMods) {
+	protoSpecified := false
+	if strings.HasPrefix(source, "file://") && len(source) > 7 {
+		protoSpecified = true
+	}
+
+	// if source, minus potential file::// prefix, exists as is, denote
```
Comment (bparees) [inline pkg/scm/git/git.go]: file://

File: pkg/scm/git/git_test.go [hunk 22]
```diff
@@ -9,26 +12,105 @@ import (
...
+		"git@192.168.122.1:repositories/authooks",
+		"mbalazs@build.ulx.hu:/var/git/eap-ulx.git",
+		"ssh://git@[2001:db8::1]/repository.git",
+		"ssh://git@mydomain.com:8080/foo/bar",
+		"git@[2001:db8::1]:repository.git",
+		"git@[2001:db8::1]:/repository.git",
+		"g_m@foo.bar:foo/bar",
+		"g-m@foo.bar:foo/bar",
+		"g.m@foo.bar:foo/bar",
+		"github.com:openshit/origin.git",
```
Comment (bparees) [inline pkg/scm/git/git_test.go]: openshift.  (don't worry, we've all done it)

File: pkg/scm/git/git_test.go [hunk 23]
```diff
@@ -9,26 +12,105 @@ import (
...
+		t.Errorf("The %q directory is not git repository", d)
+	}
+}
+
 func TestValidCloneSpec(t *testing.T) {
-	scenarios := []string{"git@github.com:user/repo.git",
+	gitLocalDir := createLocalGitDirectory(t)
+	defer os.RemoveAll(gitLocalDir)
+
+	valid := []string{"git@github.com:user/repo.git",
```
Comment (bparees) [inline pkg/scm/git/git_test.go]: needs some tests for values with # refs.

**Comment by bparees** [comment]
i think the refactor made this much more readable and easier to reason about...  good suggestion @csrwng, nice implementation @gabemontero

**Comment by gabemontero** [comment]
i totally agree on the readability, etc. improvement - thx @csrwng for the suggestion - looking at the remaining comments now

File: pkg/scm/git/git.go (hunk 20 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: yep - part of next push

File: pkg/scm/git/git.go (hunk 21 above)
Comment (gabemontero) [inline pkg/scm/git/git.go]: part of next push

File: pkg/scm/git/git_test.go (hunk 22 above)
Comment (gabemontero) [inline pkg/scm/git/git_test.go]: :-)

File: pkg/scm/git/git_test.go (hunk 23 above)
Comment (gabemontero) [inline pkg/scm/git/git_test.go]: i've added a couple

**Comment by gabemontero** [comment]
@bparees @csrwng next round - ptal

**Comment by bparees** [comment]
lgtm, will let @csrwng give it a once over monday before we merge.

**Comment by gabemontero** [comment]
Cool deal sounds good

File: pkg/scm/git/git.go [hunk 24]
```diff
@@ -34,7 +38,25 @@ type stiGit struct {
 	runner util.CommandRunner
 }
 
-var gitSshURLExp = regexp.MustCompile(`\A([\w\d\-_\.+]+@[\w\d\-_\.+]+:[\w\d\-_\.+%/]+\.git)$`)
+type URLMods struct {
```
Comment (csrwng) [inline pkg/scm/git/git.go]: godoc (otherwise make private)

File: pkg/scm/git/git.go [hunk 25]
```diff
@@ -34,7 +38,25 @@ type stiGit struct {
...
+	scheme string
+	user   string
+	host   string
+	path   string
+	ref    string
+}
+
+type FileProtoDetails struct {
```
Comment (csrwng) [inline pkg/scm/git/git.go]: godoc (otherwise make private)

**Comment by csrwng** [comment]
really minor comments, otherwise LGTM

**Comment by gabemontero** [comment]
@csrwng godoc comments added (also capitalized fields in the case where struct was public but fields were not)

**Comment by bparees** [comment]
[merge]
